server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java

```diff
@@ -242,8 +242,8 @@ public boolean decRef() {
         }
 
         @Override
-        public boolean hasReferences() {
-            return true;
+        public int refCount() {
+            return 1;
```

Comment:
The point of `hasReferences()` is to avoid having to return a fake value like this for implementations that always leak. There's something weird about having a refcount which `incRef()` and `decRef()` don't affect.

Do we need to expose the exact refcount here?
Can we put

Reply:
> is to avoid having to return a fake value like this for implementations that always leak

Right. Unfortunately, this would mean removing `org.elasticsearch.common.bytes.ReleasableBytesReference#refCount` which is quite the large change (although we eventually only make use of the number in tests seemingly) and doesn't really seem worth the effort to avoid a cosmetic issue like this? 

I guess Netty does the same and uses fake `1` for the ref count in `EmptyByteBuf` and the like so I figured it's good enough for us here as well :)

Reply:
Ah I see. Still, half of those usages are `assert refCount() > 0;` which just become `assert hasReferences();` and the others in tests are pretty much all equally trivial. I almost made that change myself when introducing `hasReferences()`. I'd rather do the Right Thing here.

Reply:
Hmm ok, let me try

Reply:
Ok I concede that this was much easier than expected. https://github.com/elastic/elasticsearch/pull/81096/commits/3cf250c3616678d0f017fef84b1bdbae2717170d ... effectively only a single spot where we lost some trivial coverage (ref count == 2 assertion)